bpf/bpf_overlay.c

```diff
@@ -299,14 +305,25 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx, __u32 *identity)
 	/* Lookup IPv4 address in list of local endpoints */
 	ep = lookup_ip4_endpoint(ip4);
 	if (ep) {
+		bool deliver_via_kernel = false;
+
 		/* Let through packets to the node-ip so they are processed by
 		 * the local ip stack.
 		 */
 		if (ep->flags & ENDPOINT_F_HOST)
 			goto to_host;
 
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE && \
+	!defined(TUNNEL_MODE) && !defined(ENABLE_HOST_ROUTING)
+		/* In conjunction with the direct routing mode, DSR with Geneve dispatch
+		 * and the legacy routing mode, we pass packets to the kernel stack.
+		 * Because the kernel stack is responsible for the routing,
+		 * e.g. CNI chain setup
+		 */
+		deliver_via_kernel = true;
+#endif
```

Comment:
In this case, where/when would we apply the endpoint's ingress policy?

Reply:
As far as I know, we apply the policy as follows.

- Pass to the stack from bpf_host(from-netdev), and the kernel stack forwards the packets to cilium_host(bpf_host from-host)
- [ipv[4,6]_local_delivery](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/lib/l3.h#L60) on bpf_host handles packets to local endpoints. 
  - In the case enable-endpoint-routes=true, call redirect_ep and call [the egress program](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L2082) loaded on the veth(to-container), and apply the policy [here](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L1552)
  - In the case enable-endpoint-routes=false, call tail_call_dynamic and apply the policy [here](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L1986)

```
# The kernel stack forwards the packets to the local endpoints to cilium_host
vagrant@k8s1:~$ ip r
default via 10.0.2.2 dev enp0s3 proto dhcp src 10.0.2.15 metric 100
10.0.0.0/24 via 10.0.0.64 dev cilium_host src 10.0.0.64
10.0.0.64 dev cilium_host scope link
vagrant@k8s1:~$ kubectl get po -o wide
NAME                        READY   STATUS    RESTARTS   AGE   IP           NODE   NOMINATED NODE   READINESS GATES
testds-tlvfm                2/2     Running   0          12m   10.0.0.119   k8s1   <none>           <none>
testds-xcx65                2/2     Running   0          12m   10.0.1.154   k8s2   <none>           <none>
vagrant@k8s1:~$ ip r get 10.0.0.119
10.0.0.119 dev cilium_host src 10.0.0.64 uid 1000
    cache

# The egress programs(to-container) are attached if endpoint route is enabled
tc:
cilium_net(50) clsact/ingress cil_to_host-cilium_net id 7196
cilium_host(51) clsact/ingress cil_to_host-cilium_host id 7176
cilium_host(51) clsact/egress cil_from_host-cilium_host id 7183
lxc_health(116) clsact/ingress cil_from_container-lxc_health id 7283
lxc_health(116) clsact/egress cil_to_container-lxc_health id 7279
lxc1151b03359c3(118) clsact/ingress cil_from_container-lxc1151b03359c3 id 7326
lxc1151b03359c3(118) clsact/egress cil_to_container-lxc1151b03359c3 id 7310
lxc148700bf712d(120) clsact/ingress cil_from_container-lxc148700bf712d id 7291
lxc148700bf712d(120) clsact/egress cil_to_container-lxc148700bf712d id 7330
```

In the generic-veth chain setup, the chained CNI is responsible for the host routing (normally the chained plugin setup the kernel FIB). The kernel stack forwards packets to local endpoints to the veth and Cilium applies the policy on the to-container program which is the same as the endpoint route is enabled case.